File: khs/자료구조/에디터.cpp
#include <iostream>
#include <string>
using namespace std;

class Node{
    public:
        char alpha;
        Node* lft;
        Node* rgt;

        Node(char a = '0'){      
            this->alpha = a;
            lft = nullptr;
            rgt = nullptr;
    }   

    Node* add_right(char a){
        this->rgt = new Node(a);
        this->rgt->lft = this;
        return this->rgt;
    }

    Node* get_left(){
        return this->lft;
    }

    Node* move_left(){
        return this->lft;
    }

    Node* get_right(){
        return this->rgt;
    }

    Node* move_right(){
        return this->rgt;
    }

    char get_alpha(){
        return this->alpha;
    }

        
};


int main(){

    string str;
    string order;
    char a;
    cin >> str;
    int M;
    cin >> M;


    Node* cursor = new Node();
    Node* start = cursor;
    Node* temp;

    for (auto c = str.begin(); c < str.end(); c++){
        cursor = cursor->add_right(*c);
    }


    while(M--){
        
        cin >> order;
        if (order == "P"){
            cin >> a;
        }
        if (order[0] == 'L'){
            if (cursor->get_left() != nullptr)
                cursor = cursor->move_left();
        }

        if (order[0] == 'D'){
            if (cursor->get_right() != nullptr)
                cursor = cursor->move_right();
        }

        if (order[0] == 'B'){
            if (cursor->get_alpha() != '0'){

                temp = cursor;
                cursor->lft->rgt = cursor->rgt;
                if (cursor->rgt)
                    cursor->rgt->lft = cursor->lft;
                cursor = cursor->lft;
                temp->lft = nullptr;
                temp->rgt = nullptr;
                delete temp;
            }

        }

        if (order[0] == 'P'){
         

            Node* c = new Node(a);
            if (cursor->rgt == nullptr){
                c->lft = cursor;
                cursor->rgt = c;
                cursor = c;
                continue;
            }
            else{
                c->rgt = cursor->rgt;
                c->lft = cursor;
                c->lft->rgt = c;
                c->rgt->lft = c;
                cursor = c;
            }

        }
    }


    while(1){
        start = start->rgt;
        cout << start->alpha;
        if (start->rgt == nullptr)
            break;
    }  


}

File: khs/자료구조/에디터.py
string = str(input())

M = int(input())

class String:
    def __init__(self, alpha="0"):
        self.alpha = alpha
        self.lft = None
        self.rgt = None
    
    def add_right(self, alpha):
        self.rgt = String(alpha)
        self.rgt.lft = self
        return self.rgt


string = list(string)

cursor = String()
start = cursor

for c in string:
    cursor = cursor.add_right(c)
# print(cursor.alpha)
# print(start.alpha)
# while True:
#     start = start.rgt
#     print(start.alpha, end="")
    
#     if start.rgt == None:
#         break
# print()
# while True:
    
#     print(cursor.alpha, end="")
#     cursor = cursor.lft
#     if cursor.lft == None:
#         break
for _ in range(M):
    order = list(map(str , input().split()))
    if order[0] == 'L':
        if cursor.lft != None: # none 대체
            cursor = cursor.lft
 
    if order[0] == 'D':
        if cursor.rgt != None:
            cursor = cursor.rgt

    if order[0] == 'B':
        if cursor.alpha != "0":
            temp = cursor
     
            cursor.lft.rgt = cursor.rgt
            if cursor.rgt:
                cursor.rgt.lft = cursor.lft
            cursor = cursor.lft
            temp.lft = None
            temp.rgt = None         
        
    if order[0] == 'P':
        c = String(order[1])
        if cursor.rgt == None:
            c.lft = cursor
            cursor.rgt = c
            cursor = c 
            continue
        else:
            c.rgt = cursor.rgt
            c.lft = cursor
            c.lft.rgt = c
            c.rgt.lft = c
            cursor = c
        
while True:
    start = start.rgt
    print(start.alpha, end="")
    
    if start.rgt == None:
        break